lib/stm32/h7/rcc.c

```diff
@@ -6,16 +6,333 @@
  *
  * LGPL License Terms @ref lgpl_license
  */
-
+#include <stddef.h>
 #include <libopencm3/cm3/assert.h>
 #include <libopencm3/stm32/rcc.h>
 #include <libopencm3/stm32/pwr.h>
 #include <libopencm3/stm32/flash.h>
 
-/**@{*/
+#define PLL1_INDEX			0
+#define PLL2_INDEX			1
+#define PLL3_INDEX			2
+#define HZ_PER_MHZ		 	1000000UL
+#define HZ_PER_KHZ			1000UL
+
+/* Local private copy of the clock configuration for providing user with clock tree data. */
+static struct {
+	uint16_t sysclk_mhz;
+	uint16_t cpu_mhz;
+	uint16_t hclk_mhz;
+	struct {
+		uint16_t pclk1_mhz;		/* APB1 clock. */
+		uint16_t pclk2_mhz;		/* APB2 clock. */
+		uint16_t pclk3_mhz;		/* APB3 clock. */
+		uint16_t pclk4_mhz;		/* APB4 clock. */
+	} per;
+	struct {					/* Each PLL output set of data. */
+		uint16_t p_mhz;
+		uint16_t q_mhz;
+		uint16_t r_mhz;
+	} pll[3];
+	uint16_t hse_khz;			/* This can't exceed 50MHz */
+} rcc_clock_tree = {
+	.sysclk_mhz = HSI_BASE_FREQUENCY / HZ_PER_MHZ,
+	.cpu_mhz = HSI_BASE_FREQUENCY / HZ_PER_MHZ,
+	.hclk_mhz = HSI_BASE_FREQUENCY / HZ_PER_MHZ,
+	.per.pclk1_mhz = HSI_BASE_FREQUENCY / HZ_PER_MHZ,
+	.per.pclk2_mhz = HSI_BASE_FREQUENCY / HZ_PER_MHZ,
+	.per.pclk3_mhz = HSI_BASE_FREQUENCY / HZ_PER_MHZ,
+	.per.pclk4_mhz = HSI_BASE_FREQUENCY / HZ_PER_MHZ
+};
```

Comment:
this is for your "get_clock[rate]" stuff right?  this sort of thing _must_ be considered for everyone else too, if at all.

Reply:
Not sure what to do with this.  This was thinking about everyone else in that they don't need to know the clock tree to fetch a clock frequency.